sweep/sweeper.go

```diff
@@ -30,13 +74,583 @@ type UtxoSweeperConfig struct {
+	_, currentHeight, err := s.cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return false, err
+	}
+
+	// Filter for inputs that need to be swept and have a positive net
+	// yield.
+	var sweepableInputs []inputYield
+	for _, input := range s.pendingInputs {
+		// Skip inputs that are already part of a sweep tx.
+		if input.state != stateNew {
+			continue
+		}
+
+		addedFee := satPerKW.FeeForWeight(int64(input.addedWeight))
+		netYield := input.input.SignDesc().Output.Value -
+			int64(addedFee)
+
+		// Skip inputs that have a negative net yield. Leave them
+		// pending until fees come down.
+		if netYield <= 0 {
+			continue
+		}
+
+		sweepableInputs = append(sweepableInputs, inputYield{
+			input:    input,
+			netYield: netYield,
+		})
+	}
+
+	// Sort input by added fee. We will start constructing sweep
+	// transactions starting with the highest yield inputs. This is to
+	// prevent the construction of a sweep tx with an output below the dust
+	// limit, causing the sweep process to stop, while there are still
+	// higher yield inputs sweepable.
+	sort.Slice(sweepableInputs, func(i, j int) bool {
+		return sweepableInputs[i].netYield > sweepableInputs[j].netYield
+	})
+
+	// Select block of inputs up to the configured maximum number.
+	for len(sweepableInputs) > 0 {
+		sweepCount := len(sweepableInputs)
+		if sweepCount > s.cfg.MaxInputsPerTx {
+			sweepCount = s.cfg.MaxInputsPerTx
+		}
+
+		// If fullOnly, don't sweep if less inputs than the maximum.
+		if fullOnly && sweepCount < s.cfg.MaxInputsPerTx {
+			return false, nil
+		}
+
+		var weightEstimate lnwallet.TxWeightEstimator
+		weightEstimate.AddP2WKHOutput()
+		weight := weightEstimate.Weight()
+
+		inputList := make([]Input, sweepCount)
+		var total int64
+		for idx, input := range sweepableInputs[:sweepCount] {
+			inputList[idx] = input.input.input
+			weight += input.input.addedWeight
+			total += input.input.input.SignDesc().Output.Value
+		}
+
+		fee := satPerKW.FeeForWeight(int64(weight))
+		outputValue := total - int64(fee)
+
+		// If the output value of this block of inputs does not reach
+		// the dust limit, stop sweeping. Because of the sorting,
+		// continuing with remaining inputs will only lead to
+		// transactions with a even lower output value.
+		if outputValue < int64(s.dustLimit) {
+			return false, nil
+		}
+
+		s.sweep(inputList, uint32(currentHeight), int64(fee))
+
+		sweepableInputs = sweepableInputs[sweepCount:]
+	}
+
+	// Disable countdown if it was active. This prevents two sweeps from
+	// happening in fast succession. The first sweep triggered by the
+	// maximum allowed number of inputs and the second because the countdown
+	// reaches 0.
+	s.sweepCountdown = 0
+
+	return true, nil
+}
+
+func (s *UtxoSweeper) sweep(inputList []Input, currentHeight uint32,
+	fee int64) {
+
+	sweepTx, err := s.createAndPublishTx(inputList, currentHeight, fee)
+	var sweepTxHash chainhash.Hash
+	if err != nil {
+		log.Error(err)
```

Comment:
I think we want to, because it shouldn't happen anymore now that we have an accurate overview of all our outstandig txes. Only in case the remote party spend and we haven't processed the spend ntfn, this can still happen.

Reply:
Exactly